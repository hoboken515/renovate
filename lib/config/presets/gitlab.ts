import { PLATFORM_FAILURE } from '../../constants/error-messages';
import { logger } from '../../logger';
import { api } from '../../platform/gitlab/gl-got-wrapper';
import { ensureTrailingSlash } from '../../util/url';
import { Preset } from './common';

const { get: glGot } = api;

export async function getDefaultBranchName(
  urlEncodedPkgName: string,
  endpoint: string
): Promise<string> {
  const branchesUrl = `${endpoint}projects/${urlEncodedPkgName}/repository/branches`;
  type GlBranch = {
    default: boolean;
    name: string;
  }[];

  const res = await glGot<GlBranch>(branchesUrl);
  const branches = res.body;
  let defautlBranchName = 'master';
  for (const branch of branches) {
    if (branch.default) {
      defautlBranchName = branch.name;
      break;
    }
  }

  return defautlBranchName;
}

export async function fetchJSONFile(
  repo: string,
  fileName: string,
  endpoint: string
): Promise<Preset> {
  const urlEncodedPkgName = encodeURIComponent(repo);
  const defaultBranchName = await getDefaultBranchName(
    urlEncodedPkgName,
    endpoint
  );
  const url = `${endpoint}projects/${urlEncodedPkgName}/repository/files/${fileName}?ref=${defaultBranchName}`;
  let res: { body: { content: string } };
  try {
    res = await glGot(url);
  } catch (err) {
    if (err.message === PLATFORM_FAILURE) {
      throw err;
    }
    logger.debug(
      { statusCode: err.statusCode },
      `Failed to retrieve ${fileName} from repo`
    );
    throw new Error('dep not found');
  }
  try {
    const content = Buffer.from(res.body.content, 'base64').toString();
    const parsed = JSON.parse(content);
    return parsed;
  } catch (err) {
    throw new Error('invalid preset JSON');
  }
}

export async function getPresetFromEndpoint(
  pkgName: string,
  filePreset: string,
  endpoint = 'https://gitlab.com/api/v4/'
): Promise<Preset> {
  // eslint-disable-next-line no-param-reassign
  endpoint = ensureTrailingSlash(endpoint);
  const [fileName, presetName, subPresetName] = filePreset.split('/');
  let jsonContent;
  if (fileName === 'default') {
    try {
      jsonContent = await fetchJSONFile(pkgName, 'default.json', endpoint);
    } catch (err) {
      if (err.message !== 'dep not found') {
        throw err;
      }
      logger.debug('default.json preset not found - trying renovate.json');
      jsonContent = await fetchJSONFile(pkgName, 'renovate.json', endpoint);
    }
  } else {
    jsonContent = await fetchJSONFile(pkgName, `${fileName}.json`, endpoint);
  }
  if (presetName) {
    if (subPresetName) {
      return jsonContent[presetName]
        ? jsonContent[presetName][subPresetName]
        : undefined;
    }
    return jsonContent[presetName];
  }
  return jsonContent;
}

export async function getPreset(
  pkgName: string,
  presetName = 'default'
): Promise<Preset> {
  return getPresetFromEndpoint(pkgName, presetName);
}
