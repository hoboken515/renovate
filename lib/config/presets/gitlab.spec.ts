import { PLATFORM_FAILURE } from '../../constants/error-messages';
import { GotResponse } from '../../platform';
import { api } from '../../platform/gitlab/gl-got-wrapper';
import { clearRepoCache } from '../../util/cache';
import * as gitlab from './gitlab';
import { PartialDeep } from 'type-fest';

jest.mock('../../platform/gitlab/gl-got-wrapper');
jest.mock('../../util/got');

const glGot: jest.Mock<Promise<PartialDeep<GotResponse>>> = api.get as never;
const branchesResponse = {
  body: [
    {
      name: 'devel',
    },
    {
      name: 'master',
      default: true,
    },
  ],
};

describe('config/presets/gitlab', () => {
  beforeEach(() => {
    glGot.mockReset();
    glGot.mockResolvedValueOnce(branchesResponse);
    return global.renovateCache.rmAll();
  });
  describe('fetchJSONFile()', () => {
    beforeEach(() => {
      clearRepoCache();
    });
    it('returns JSON', async () => {
      glGot.mockResolvedValue({
        body: {
          content: Buffer.from('{"from":"api"}').toString('base64'),
        },
      });
      const res = await gitlab.fetchJSONFile(
        'some/repo',
        'some-filename',
        'https://gitlab.example.org/api/v4'
      );
      expect(res).toMatchSnapshot();
    });
  });
  describe('getPreset()', () => {
    it('passes up platform-failure', async () => {
      glGot.mockImplementation(() => {
        throw new Error(PLATFORM_FAILURE);
      });
      await expect(gitlab.getPreset('some/repo')).rejects.toThrow(
        PLATFORM_FAILURE
      );
    });
    it('tries default then renovate', async () => {
      glGot.mockImplementation(() => {
        throw new Error();
      });
      await expect(gitlab.getPreset('some/repo')).rejects.toThrow();
    });
    it('throws if no content', async () => {
      glGot.mockResolvedValue({
        body: {},
      });
      await expect(gitlab.getPreset('some/repo')).rejects.toThrow();
    });
    it('throws if fails to parse', async () => {
      glGot.mockResolvedValue({
        body: {
          content: Buffer.from('not json').toString('base64'),
        },
      });
      await expect(gitlab.getPreset('some/repo')).rejects.toThrow();
    });
    it('should return default.json', async () => {
      glGot.mockResolvedValue({
        body: {
          content: Buffer.from('{"foo":"bar"}').toString('base64'),
        },
      });
      const content = await gitlab.getPreset('some/repo');
      expect(content).toEqual({ foo: 'bar' });
    });
    it('should query preset within the file', async () => {
      glGot.mockResolvedValue({
        body: {
          content: Buffer.from('{"somename":{"foo":"bar"}}').toString('base64'),
        },
      });
      const content = await gitlab.getPreset('some/repo', 'somefile/somename');
      expect(content).toEqual({ foo: 'bar' });
    });
    it('should query subpreset', async () => {
      glGot
        .mockResolvedValueOnce({
          body: {
            content: Buffer.from(
              '{"somename":{"somesubname":{"foo":"bar"}}}'
            ).toString('base64'),
          },
        })
        .mockResolvedValueOnce(branchesResponse)
        .mockResolvedValueOnce({
          body: {
            content: Buffer.from(
              '{"somename":{"somesubname":{"foo":"bar"}}}'
            ).toString('base64'),
          },
        });
      let content = await gitlab.getPreset(
        'some/repo',
        'somefile/somename/somesubname'
      );
      expect(content).toEqual({ foo: 'bar' });
      content = await gitlab.getPreset(
        'some/repo',
        'somefile/wrongname/somesubname'
      );
      expect(content).toBeUndefined();
    });
    it('should return custom.json', async () => {
      glGot.mockResolvedValue({
        body: {
          content: Buffer.from('{"foo":"bar"}').toString('base64'),
        },
      });
      try {
        global.appMode = true;
        const content = await gitlab.getPreset('some/repo', 'custom');
        expect(content).toEqual({ foo: 'bar' });
      } finally {
        delete global.appMode;
      }
    });
    it('should return the preset', async () => {
      glGot.mockResolvedValue({
        body: {
          content: Buffer.from('{"foo":"bar"}').toString('base64'),
        },
      });
      const content = await gitlab.getPreset('some/repo');
      expect(content).toEqual({ foo: 'bar' });
    });
  });
  describe('getPresetFromEndpoint()', () => {
    it('uses custom endpoint', async () => {
      await gitlab
        .getPresetFromEndpoint(
          'some/repo',
          'default',
          'https://gitlab.example.org/api/v4'
        )
        .catch((_) => {});
      expect(glGot.mock.calls[0][0]).toEqual(
        'https://gitlab.example.org/api/v4/projects/some%2Frepo/repository/branches'
      );
    });
  });
});
