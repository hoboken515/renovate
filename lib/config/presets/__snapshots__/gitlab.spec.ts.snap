// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config/presets/gitlab fetchJSONFile() returns JSON 1`] = `
Object {
  "from": "api",
}
`;
